### ISSUE TITLE
Feature Request: New checksum algorithms in ZFS

### ISSUE THREAD

**Issue by ronnyegner** [zfsonlinux/zfs#3770]
Hi,

can we also have the new checksum algorithms proposed here:

http://lists.open-zfs.org/pipermail/developer/2015-June/001509.html

Thanks

**Comment by kernelOfTruth** [comment]
Haven't seen it turn up at illumos-gate yet

https://www.illumos.org/issues/4185 add new cryptographic checksums to ZFS: SHA-512, Skein, Edon-R
https://reviews.csiden.org/r/223/ 4185 add new cryptographic checksums to ZFS: SHA-512, Skein, Edon-R

**Comment by kernelOfTruth** [comment]
https://github.com/illumos/illumos-gate/commit/45818ee124adeaaf947698996b4f4c722afc6d1f

**Comment by hkratz** [comment]
Anyone working on that? The bad sha256 performance cripples deduplication but it also limits write speed to a fraction of what it could be whenever sha256 is enabled on a pool. And fletcher4 is not an alternative if one values verifiable data consistency

**Comment by behlendorf** [comment]
@hkratz no to my knowledge, it would be great if you could port the Illumos commit, https://github.com/illumos/illumos-gate/commit/45818ee and open a pull request.

**Comment by hkratz** [comment]
That commit touches over 100 files. It might be a bit much for me to tackle as I haven't worked on zfs at all before. Also it seems that they are using kernel routines for sha2, so it would not help with the performance of existing pools using sha256.

I could look into porting Dr. Gladman's pure C SHA2 code[1] which should be a lot faster. Is his license[2] acceptable for ZFS?

[1] http://www.gladman.me.uk/
[2] https://gist.github.com/hkratz/a85b3abe32791cac2df0

**Comment by lundman** [comment]
If it helps, we already pulled it in
https://github.com/openzfsonosx/zfs/commit/cc54f70bda0430a6267598f010b2e56e05e58a09
https://github.com/openzfsonosx/spl/commit/97657bb7b0e3d0fe00192d4b53e9cbe900c77468

But tweaking the filenames isn't too hard :)

**Comment by kernelOfTruth** [comment]
@lundman it definitely helps,

the more information the better ;)

My main question with those changes is where the crypto folder (and the subsequent files) is supposed to land

currently we have e.g. module/zfs/sha256.c for the sha256 hash

and I'm certain it would be quite some effort to split that change up into spl and zfs

**Comment by behlendorf** [comment]
We'll just need to make some decision.  Since zfs is going to be the only consumer we could add it to the zfs directory.

> and I'm certain it would be quite some effort to split that change up into spl and zfs

We're slowly working towards merging the two in to a single source tree so lets avoid adding any more code to the spl if at all possible.

**Comment by tonyhutter** [comment]
@behlendorf I can take a stab at porting this one.  @lundman thanks for the heads-up on the OSX port!  Should make life a lot easier.

**Comment by ronnyegner** [comment]
@tonyhutter I can´t help you with the programming, but i can help with testing...

**Comment by lundman** [comment]
Although, now I'm thinking about redoing it, since the crypto PR adds the whole IllumOS crypto to ZFS, and skein/edon-r just slot into that, I should remove it from SPL and move it to ZFS.

 So could be worth keeping in mind, if to base it on crypto :)

**Comment by behlendorf** [comment]
I haven't looked carefully at the dependencies needed for this patch.  But I'd be OK with applying the first half of the encryption patches which add the Illumos Crypto Framework if that functionality is needed.  I'd far prefer that to adding additional interfaces to the SPL.

**Comment by tonyhutter** [comment]
@ronnyegner thanks for the offer. I'll give you a heads-up when I'm ready to test.   @behlendorf I'll port this to ZFS only (no SPL).

**Comment by tonyhutter** [comment]
I talked with @behlendorf offline about this, and the current plan is for us to port it on top of the Crypto Framework from https://github.com/zfsonlinux/zfs/pull/4329

**Comment by tonyhutter** [comment]
Quick update:
I have a pull request up (https://github.com/zfsonlinux/zfs/pull/4760).  Unfortunately, I'm running into some stack frame size issues on ARM32, x86-32, and PPC.  I'm looking into how to resolve those in the the least ugly way possible...  Might be a similar problem to what's being reported in https://github.com/zfsonlinux/zfs/issues/4799.

**Comment by behlendorf** [comment]
The #4760 PR is in good shape according to the automated testing and I think it's at the point where it would benefit from some additional testing particularly on non-x86_64 and 32-bit platforms.

For anyone interested in getting this finalized it would be helpful to have some independent testing results from a **non-production** pool.  @ronnyegner @hkratz you commented above you're interested in this feature.  @gordan-bobic can I twist your arm to give this a spin on a arm system.

**Comment by ronnyegner** [comment]
@behlendorf I can help you.... But only on x64 or x32.... If you want me to then  i need a quick help how to get this running.

**Comment by tonyhutter** [comment]
@ronnyegner thanks for the help.  Basically, just set the checksum on a pool to one of the new checksum algorithms, write some data to a file in the pool, export the pool, import it, and look for zero checksum errors (`zpool status`).  You can set the checksum with:

`zfs set checksum=[noparity|skein|edonr|sha512] POOL`

If you feeling ambitious, do the same steps but scramble the data on the underlying vdev before importing it back again.  If everything works right, you should see checksum errors after import.  This will corrupt your data (obviously), so only try it on a test pool.

**Comment by ronnyegner** [comment]
@tonyhutter  That´s fine tony. My question was more related to how do i get the required patches since they are not yet in the master branch.

**Comment by tonyhutter** [comment]
@ronnyegner this will checkout my repo and switch to the branch with the patch:
```
git clone https://github.com/tonyhutter/zfs.git
cd zfs
git branch --track openzfs-4185 remotes/origin/openzfs-4185
git checkout openzfs-4185
```

**Comment by ronnyegner** [comment]
@tonyhutter 

I tried to compile the stack (with SPL from your repo) and i get the following error during compile:

  CC [M]  /root/zfs/module/icp/os/modhash.o
/root/zfs/module/icp/os/modhash.c: In function ‘mod_hash_create_ptrhash’:
/root/zfs/module/icp/os/modhash.c:315:2: error: implicit declaration of function ‘highbit’ [-Werror=implicit-function-declaration]
  rshift = highbit(key_elem_size);
  ^
cc1: some warnings being treated as errors
make[5]: *** [/root/zfs/module/icp/os/modhash.o] Error 1
make[4]: *** [/root/zfs/module/icp] Error 2
make[3]: *** [_module_/root/zfs/module] Error 2
make[3]: Leaving directory `/usr/src/linux-headers-3.14.43-031443-generic'
make[2]: *** [modules] Error 2
make[2]: Leaving directory `/root/zfs/module'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/root/zfs'


System is a Ubuntu 14.04. x86_64 with 3.14.43-031443-generic Kernel.


Ronny

**Comment by behlendorf** [comment]
@ronnyegner you'll want to use the SPL from the master branch.  Specifically, you'll want a version which contains this commit d2f97b2a2651d8e1a6e9e1dcb07cfe8570efcfff.  Thanks for the help!

**Comment by ronnyegner** [comment]
@behlendorf  I think i have checked out the current master and compiled it by doing:

`svn co https://github.com/zfsonlinux/spl.git`

The error is still the same.

**Comment by behlendorf** [comment]
Resolved by 3c67d83a8afb391f20bc53d36a0cebea6897b3e2.
